### ISSUE TITLE
Cannot install bisect_ppx

### ISSUE THREAD

**Issue by Toxaris** [pnpm/pnpm#3817]
### pnpm version:

2.6.3

### Code to reproduce the issue:

```
pnpm add bisect_ppx
```

### Expected behavior:

[bixect_ppx](https://www.npmjs.com/package/bisect_ppx) is installed
### Actual behavior:

```
ENOENT: no such file or directory, chmod '.../node_modules/.pnpm/bisect_ppx@2.6.3/node_modules/bisect_ppx/bisect-ppx-report'
```

### Additional information:

I'm testing this in a docker container on Docker for Desktop with the WSL2 backend on Windows.

```
$ uname -a
Linux db71fafc113a 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 GNU/Linux

$ node -v
v12.22.6

$ pnpm -v
6.16.0
```

**Comment by Toxaris** [comment]
I looked into the details a bit and I think the problem is this:

- `bisect_ppx` bundles binaries for common operating systems in subfolders
- a `preinstall` script detects the operating system and copies the correct file to the root of the package
- a `bin` entry points to the file name in the root of the package

It appears that `pnpm` tries to `chmod` the file in the root of the package before it has been copied there.

I guess the `chmod` step should move to after the `preinstall`, `install`,  and `postinstall` lifecycle events?

**Comment by wre232114** [comment]
preinstall, install, postinstall are all executed after the real installation, and bin may be used in these scripts so they need to be created before the execution, see #3760, #3763

**Comment by Toxaris** [comment]
Indeed, it appears that `bisect_ppx` does not work with `yarn` either, for the same reason, see [yarn#3421](https://github.com/yarnpkg/yarn/issues/3421#issuecomment-602218701) and [bisect_ppx#290](https://github.com/aantron/bisect_ppx/issues/290).

I opened [bisect_ppx#381](https://github.com/aantron/bisect_ppx/issues/381) about my issue with pnpm and bisect_ppx.

> I think bisect_ppx may need to be updated to work with the latest pnpm/yarn/npm

Is there a recommended approach for generating a `bin` file during the package installation that works across pnpm/yarn/npm?

**Comment by Toxaris** [comment]
I discovered that we are not actually using the `bin` of bisect_ppx in our project, so I put the following work-around in the `.pnpmfile.cjs`:

```js
if (name === "bisect_ppx") {
  context.log(`${name}: removing bin`);
  delete packageJson.bin;
}
```

**Comment by aantron** [comment]
@wre232114, reading https://github.com/npm/cli/issues/2660, I have the (perhaps wrong) impression that npm intends to revert to running `preinstall` before the actual installation. If that's right, can Bisect continue to rely on that in the long term?

I'm not aware of any other way of creating the right binary at the location expected by pnpm before it attempts to do `chmod`. I could probably put a dummy binary there, and then read its permissions, overwrite it in the `preinstall` step, and apply the permissions to the final binary.

**Comment by zkochan** [comment]
This seems to work with pnpm v8.6.5

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm add bisect_ppx` installs 2.8.2 and links bisect-ppx-report into node_modules/.bin, with no ENOENT on chmod.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
